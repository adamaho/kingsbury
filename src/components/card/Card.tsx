import * as React from 'react';

import styled, {
  css
} from 'styled-components';

import {
  theme
} from '../../theme';

import {
  Header,
  HeaderProps
} from './Header';

import {
  Footer,
  FooterProps
} from './Footer';

interface CardFunctionComponent<T> extends React.FunctionComponent<T> {
  Header: React.FunctionComponent<HeaderProps>;
  Footer: React.FunctionComponent<FooterProps>;
}

interface CardProps {
  /** Content to show in the card */
  children?: React.ReactNode;

  /** classname for the card */
  className?: string;

  /** Function to handle click event */
  onClick?: () => void;

  /** Global theme in ThemeProvider */
  theme?: any; // eslint-disable-line
}

const Container = styled.div`
  height: ${(props) => props.theme.card.height};
  width: ${(props) => props.theme.card.width};

  background: ${(props) => props.theme.card.background};
  border: ${(props) => props.theme.card.outerBorder};
  box-shadow: ${(props) => props.theme.card.boxShadow};
  border-radius: ${(props) => props.theme.card.borderRadius};

  opacity: ${(props) => (props.onClick ? 0.8 : 1)};

  cursor: ${(props) => (props.onClick ? 'pointer' : 'default')};

  transition: ${(props) => props.theme.animations.time.fast} cubic-bezier(.19, 1, .4, 1);

  ${(props) => props.onClick && css`
    &:hover {
      opacity: 1;
      transform: translateY(-1%);
      box-shadow: 0 7px 21px 0 rgba(0,0,0,.07);
    }
  `}
`;

export const Card: CardFunctionComponent<CardProps> = (props) => {
  const {
    className,
    children,
    onClick,
    theme
  } = props;

  return (
    <Container
      className={className}
      onClick={onClick}
      theme={theme}
      {...props}
    >
      {children}
    </Container>
  )
};

Card.Header = Header;
Card.Footer = Footer;

Card.defaultProps = {
  children: '',
  className: '',
  onClick: undefined,
  theme,
};

