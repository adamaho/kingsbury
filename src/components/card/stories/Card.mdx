import { Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { Title, SectionTitle } from '../../../../.storybook/components';
import { Card } from '../Card';

<Title
  title="Card"
  subtitle="Cards allow users to display information about a specific subject."
/>

<SectionTitle>Simple</SectionTitle>

A simple card can be created about a specific subject without a header or footer.

<Preview>
  <Story id="components-card--simple"/>
</Preview>

<SectionTitle>Complex</SectionTitle>

A complex card can be created about a specific subject simply by adding a header and footer.

<Preview>
  <Story id="components-card--complex"/>
</Preview>

<SectionTitle>Clickable</SectionTitle>

In some cases. the cards may need to be clickable to allow users to navigate to more detailed information about that subject.

<Preview>
  <Story id="components-card--clickable"/>
</Preview>

<SectionTitle>Props</SectionTitle>
<Props of={Card} />
