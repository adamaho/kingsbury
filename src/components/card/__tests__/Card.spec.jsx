import React from 'react';
import enzyme from 'enzyme';
import renderer from 'react-test-renderer';

import Card from '..';

describe('Card', () => {
  test('it renders without children', () => {
    const tree = renderer.create(<Card />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders header', () => {
    const tree = renderer.create(
      <Card>
        <Card.Header />
      </Card>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it renders footer', () => {
    const tree = renderer.create(
      <Card>
        <Card.Footer />
      </Card>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('it calls onClick', () => {
    const onClickMock = jest.fn();
    const component = enzyme.mount(
      <Card onClick={onClickMock} />
    );

    component.find(Card).simulate('click');
    expect(onClickMock).toHaveBeenCalled();
  });
});
