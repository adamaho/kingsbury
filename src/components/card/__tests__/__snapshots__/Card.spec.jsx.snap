// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card it renders footer 1`] = `
.c1 {
  background: #FFFFFF;
  height: auto;
  min-height: 50px;
  border-top: 1px solid #C4C4C4;
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 20px;
}

.c0 {
  height: auto;
  width: auto;
  background: #FFFFFF;
  border: none;
  box-shadow: 0px 0px 7px rgba(0,0,0,0.20);
  border-radius: 20px;
  opacity: 1;
  cursor: default;
  -webkit-transition: 0.25s cubic-bezier(.19,1,.4,1);
  transition: 0.25s cubic-bezier(.19,1,.4,1);
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    
  </div>
</div>
`;

exports[`Card it renders header 1`] = `
.c1 {
  background: #FFFFFF;
  border-bottom: 1px solid #C4C4C4;
  border-top-left-radius: 20px;
  border-top-right-radius: 20px;
  height: 50px;
}

.c0 {
  height: auto;
  width: auto;
  background: #FFFFFF;
  border: none;
  box-shadow: 0px 0px 7px rgba(0,0,0,0.20);
  border-radius: 20px;
  opacity: 1;
  cursor: default;
  -webkit-transition: 0.25s cubic-bezier(.19,1,.4,1);
  transition: 0.25s cubic-bezier(.19,1,.4,1);
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    
  </div>
</div>
`;

exports[`Card it renders without children 1`] = `
.c0 {
  height: auto;
  width: auto;
  background: #FFFFFF;
  border: none;
  box-shadow: 0px 0px 7px rgba(0,0,0,0.20);
  border-radius: 20px;
  opacity: 1;
  cursor: default;
  -webkit-transition: 0.25s cubic-bezier(.19,1,.4,1);
  transition: 0.25s cubic-bezier(.19,1,.4,1);
}

<div
  className="c0"
>
  
</div>
`;
